### DISCUSSION TITLE
docs/argocd/

### DISCUSSION THREAD

**Discussion by giscus** [ricsanfre/pi-cluster#437] [Announcements]
# docs/argocd/

How to apply GitOps to Pi cluster configuration using ArgoCD.

https://picluster.ricsanfre.com/docs/argocd/

**Comment by ilbarone87** [comment]
Hi @ricsanfre ,
great work on the reworking of the application repo structure and the move to Cilium (I use it since i started with kubernetes and it's great).
I have a question for you I would like to move the deployment of my ArgoCD and Longhorn from manifest to Helm Chart without risking to completely break my cluster (especially because ArgoCD manages all my applications and longhorn the pvc where i have data stored). Do you have any advice/experience with that?
Thanks

**Reply by ricsanfre** [reply]
Hi @ilbarone87, thanks for your comments

Both with the current repo structure and the previous one, I was deploying ArgoCD and Longhorn from helm charts. Before I was using the pattern "Helm Umbrella Chart", because values.yaml file could not be in the same repo as the chart, and now I am using kustomize helm chart inflation capability. You can see details of both patterns described in the argocd installation and configuration document.

However, there is a new ArgoCd functionality, currently (release v2.11.2) in beta status, to support multiple source repos per Application, that will remove the limitation behind Umbrella chart design pattern. With this functionality a Helm Chart application can be defined in ArgoCD using a `values.yaml` file located in your own repo (different from the repo containing the chart binaries).
See details in ["ArgoCD documentation: Multiple Sources for an Application"](
https://argo-cd.readthedocs.io/en/stable/user-guide/multiple_sources/).

Both options, Kustomized Application using Helm Chart Inflator, and Helm Chart Application with multiple source repos, shoud provide same results. At the end ArgoCD Helm chart application is using the same helm chart inflation mechanism as Kustomize (`helm template` command). I decide to use kustomized version because of its flexibility: capability of having different configuration for different environments (overlays) and applying patches to manifest files after the helm chart inflation. 

Regards
Ricardo

**Comment by mitchross** [comment]
Can you provide some basic steps to just leverage your repo to bootstrap argo and a few apps?  I really like your Argo layout, but its kind of hard to set up if you dont want to bootstrapk3s, pxes etc.

Basically I want to leverage your repo/ansible for 

 Argo setup, Cilium, System Upgrade controller, and idk lets say prometheus grafana...

Im ok with installing k3s myself and everything else.. I just need some help with a automated argocd install.

What are the few commands id have to run? do I have to pre make all the argo folders? ie base, overlay, infra, base, root app?

It would be super cool if you can add this functionality.. 

thanks!

**Reply by ricsanfre** [reply]
Hi @mitchross

Did you check [quick install instructions](https://picluster.ricsanfre.com/docs/ansible/) ?

If you already have a cluster available and you do not want to automate the set-up of those nodes you can skip those sections in the quick-instructions. You only need to prepare Ansible control node as explained in the instructions and configure accordingly the inventory and var files, and jump directly to K3s installation and bootstrap

For installing the k3s cluster, not including cilium CNI installation, there is already an Ansible playbook, [ansible/k3s_install.yml](https://github.com/ricsanfre/pi-cluster/blob/master/ansible/k3s_install.yml), which is automatically executed with `make k3s-install` command.

Once the K3s cluster is installed, Cilium installation, ArgoCD deployment and applications bootstrap is automated using another another Ansible playbook, [ansible/k3s_bootstrap,yml](https://github.com/ricsanfre/pi-cluster/blob/master/ansible/k3s_bootstrap.yml), which is automatically executed with `make k3s-bootstrap` command.

This playbook specify step by step what are the commands that need to be executed. Most of Ansible tasks are shell commands (`kubectl`commands) to deploy the Kustomized applications.

You can extract from the Ansible playbook the kubectl commands needed for cluster bootstraping.

If you want to install only Cilium and bootstraping the cluster, once K3s is installed, the commands to execute are listed in [dev/Makefile](https://github.com/ricsanfre/pi-cluster/blob/master/dev/Makefile). These files are new in the repo with the intention of having a dev environment using k3d.

```shell

crds-install:
	kubectl kustomize --enable-helm --load-restrictor=LoadRestrictionsNone \
	   ../kubernetes/infrastructure/crds/overlays/dev | kubectl apply --server-side -f - 

cilium-install:
	kubectl kustomize --enable-helm --load-restrictor=LoadRestrictionsNone \
	   ../kubernetes/infrastructure/cilium/overlays/dev | kubectl apply -f -

cilium-cofig:
	kubectl kustomize --enable-helm --load-restrictor=LoadRestrictionsNone \
	   ../kubernetes/infrastructure/cilium-config/overlays/dev | kubectl apply -f -


cluster-bootstrap:
	kubectl kustomize --enable-helm --load-restrictor=LoadRestrictionsNone \
	   ../kubernetes/infrastructure/argocd/overlays/dev | kubectl apply -f -
	kubectl kustomize --enable-helm --load-restrictor=LoadRestrictionsNone \
	   ../kubernetes/bootstrap/argocd/overlays/dev | kubectl apply -f -

```

The above commands are using the `dev` overlay. You can launch similar commands using `prod` overlay.


Regards
Ricardo
